### PR TITLE
Deletes recipient_reagents_holder on del

### LINKED ISSUE #57701: Plumbing components can cause /datum/reagent to harddelete
```
[2021-03-14 05:01:04.877] ## TESTING: Beginning search for references to a /datum/reagents.
[2021-03-14 05:06:18.895] ## TESTING: Found /datum/reagents [0x21012cba] in /datum/component/plumbing/simple_demand's reagents var. World -> /datum/component/plumbing/simple_demand /datum/component/plumbing/simple_demand
[2021-03-14 05:06:18.900] ## TESTING: Found /datum/reagents [0x21012cba] in /datum/component/plumbing/simple_demand's recipient_reagents_holder var. World -> /datum/component/plumbing/simple_demand /datum/component/plumbing/simple_demand
[2021-03-14 05:06:29.512] ## TESTING: Completed search for references to a /datum/reagents.
```

### PR DESCRIPTION
**Pull request by Time-Green**
Closes #57701

:cl:
fix: fixes plumbing components letting some reagent holders not properly delete
/:cl:

I think this is what they meant, I'm not entirely sure

### CODE REVIEW

**Comment by LemonInTheDark** [comment]
The reagent held by AM is being deleted, but the component is living on past the deletion, which causes harddeletes.
The question is should the component go away if this happens, or should something else happen

**Comment by Time-Green** [comment]
> The reagent held by AM is being deleted, but the component is living on past the deletion, which causes harddeletes.
> The question is should the component go away if this happens, or should something else happen

It should 100% go away, I never realized they didn’t. Aren’t components supposed to go away with their parent? I’ll look into it tomorrow to see how

**Comment by Time-Green** [comment]
Kind of forgot about this. I understand what was meant now, and when the recipient_reagents_holder (another objects reagentholder) is deleted it properly cleans it out so the other object can delete itself in peace

**Comment by LemonInTheDark** [comment]
Does this solve the `var/reagent` aspect of the hard delete as well?

**Comment by Time-Green** [comment]
> Does this solve the `var/reagent` aspect of the hard delete as well?

I don't know. I checked in-game and 'var/reagents' does get set to null and the component then gets deleted, but I'm not sure if that means there's no hard-deletion

File: code/datums/components/plumbing/_plumbing.dm [hunk 1]
```diff
@@ -42,25 +42,28 @@
...
-	if(custom_receiver)
-		recipient_reagents_holder = custom_receiver
-	else
-		recipient_reagents_holder = AM.reagents
+	set_recipient_reagents_holder(custom_receiver ? custom_receiver : AM.reagents)
 
+	if(start)
+		//timer 0 so it can finish returning initialize, after which we're added to the parent.
+		//Only then can we tell the duct next to us they can connect, because only then is the component really added. this was a fun one
+		addtimer(CALLBACK(src, .proc/enable), 0)
```
Comment (AnturK) [inline code/datums/components/plumbing/_plumbing.dm:50]: This looks like something that should be systemic or you could register to component added signal instead and unregister in response.

**Review by AnturK — CHANGES_REQUESTED** [review]

File: code/datums/components/plumbing/_plumbing.dm (hunk 1 above)
Comment (ninjanomnom) [reply code/datums/components/plumbing/_plumbing.dm:50]: Yeah sleeps, even for 0, to deal with init issues isn't the way to go. React to something else when completed.

Comment (Time-Green) [reply code/datums/components/plumbing/_plumbing.dm:50]: Do you mean just send a signal?

Comment (ninjanomnom) [reply code/datums/components/plumbing/_plumbing.dm:50]: Yeah, you could do what antur mentioned or listen for some other signal that happens around when you want it. By react though I just triggering this code by some method when other code runs.

**Comment by Time-Green** [comment]
Plumbing listents to the COMPONENT_ADDED signal to know when to enable and potentially built a network, I tested it as a bonus to confirm this doesn't kill plumbing

File: code/datums/components/plumbing/_plumbing.dm [hunk 2]
```diff
@@ -210,8 +210,8 @@
 					duct.update_appearance()
 
 ///settle wherever we are, and start behaving like a piece of plumbing
-/datum/component/plumbing/proc/enable()
-	if(active)
+/datum/component/plumbing/proc/enable(obj/object, datum/component/component)
+	if(active || (component && component != src))
 		return
```
Comment (ninjanomnom) [inline code/datums/components/plumbing/_plumbing.dm:216]: Unregister from the signal once you've been enabled

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by Time-Green** [comment]
Thanks for your patience, one of these days I'll get it right on the first try

**Review by ninjanomnom — APPROVED** [review]

**Pull request merged**